Centralize and unify network_time updating

awelzel opened the pull request:
#### Background

There are various places where network_time is forwarded to the current time, making issues like #2856 and generally network_time behavior somewhat hard to reason about. Further, external scripts, plugins or Zeek modifications that want to tightly control network_time behavior can be surprised that, for example, an incoming broker message updates network time to the current time.

#### Approach

Implement a `maybe_forward_network_time()` function and call it from the main-loop before draining the event manager unless time has been updated by processing a packet, or the packet source isn't considered idle yet. Remove all other places where network_time is updated.

Further, support completely disabling network_time forwarding via `redef allow_network_time_forward = F`. With this set, the only way to update network time would be packets from a packet source, the `set_network_time()` bif that can be called from broker events in script land, or an external plugin registering as an IO source and calling  `update_network_time()` whenever it wants to, thereby overtaking network_time updates when no packet source is available.

The consequences of this changeset that I'm aware of:
* In low traffic environments, timer expiration may be delayed by `packet_source_inactivity_timeout` (100 msec default).
  * I could actually see us going down to 20msec given that the af_packet plugin has a [10msec block expiration](https://github.com/zeek/zeek-af_packet-plugin/blob/00143ebf1720d30a355b5e5f34c4863c895201b1/scripts/init.zeek#L12-L13).
  * In other environment with larger NIC buffers and/or larger block/segment timeouts, it can be increased as needed.
 
* `suspend_processing()` in non-pseudo-realtime mode does not forward time anymore. This was mostly conscious and the test was disabled: IMO stalling `network_time` is better behavior as otherwise it is immediately forwarded to current time. See comment on #2856 from Jan. So while timers would be stuck in that mode, at least network_time behaves predictably. A few tests were added to show interaction of network_time and suspend_processing. Previously came up as buggy behavior on the [forum](https://community.zeek.org/t/potential-bug-of-network-time-when-used-together-with-suspend-processing-and-continue-processing/5086).


* An event scheduled from `zeek_init()` in `zeek -r` mode will have `network_time()` set to 0.0. This came up previously in GH-938 and lead to the introduction of `network_time_init()`, seems today we can let go of it. The processing of the first packet prevents use cases that want to send info about the first connection over broker as recently brought up in the [forum](https://community.zeek.org/t/connection-flow-not-detected-in-new-connection-but-in-connection-state-remove/6938/5).

* Mostly an observation: Except for the first `DoUpdate()` call, input readers are driven by network time. This makes using the raw reader with a command (think `while true; do sleep 0.1; echo "tick" ; done`) as a time source with `allow_network_time_forward=F` impossible. There may not be a practical use-case for this, but it would be nice if that just worked conceptually.
  * My take is that this is mostly an issue how readers rely on a network-time based heartbeat to read instead of being really async and propagating their readiness via Flares/pipes back to the main loop.
 
---

This currently contains #2866 which should go first.

J-Gras reviewed (commented):
I like the centralized approach. Seems indeed easier to reason about. I left some comments inline.

One thing we discussed out-of-band: There was some confusion about the behavior of `Poll()` when there are IO sources with 0-timeout but it is time to poll. Looks like the poll is enforced and might delay processing of the ready sources that do not provide a FD.

awelzel commented on src/broker/Manager.cc:
```diff
@@ -1140,11 +1140,6 @@ void Manager::DispatchMessage(const broker::topic& topic, broker::data msg)
 	{
-	// Ensure that time gets update before processing broker messages, or events
-	// based on them might get scheduled wrong.
-	if ( use_real_time )
```
I can't say I'm 100% sure of the implications, with this changeset network time might lag a bit behind realtime. Events/timers scheduled from broker events will now be based on Zeek's actual notion of network time based on packets, rather than forcing time to be forwarded whenever a message is received.

For broker stores today, at least in the live realtime case, expiration should happen wherever the master store is located and that may have no relation to a workers notion of network time (and fast forwarding time doesn't help a lot here).

I do think that `allow_network_time_forward=F` should  set Broker's use_real_time false, too, or there needs to be a new option to control that separately. I'll add that, but we can iterate in follow-ups if it there's a need to control that separately.

J-Gras replied:
Makes sense! I am wondering whether it would be worth to add a metric to the telemetry framework that provides the difference between network time and wall clock. Probably only relevant if there are setups that need a large idle timeout.

timwoj replied:
> For broker stores today, at least in the live realtime case, expiration should happen wherever the master store is located and that may have no relation to a workers notion of network time (and fast forwarding time doesn't help a lot here).

With all of these changes, how does master's time move forward?

awelzel replied:
> Is this worked on or tracked anywhere?

IMO easiest to add via an external script as a gauge.

awelzel commented on src/iosource/Manager.cc:
```diff
@@ -157,25 +157,24 @@ void Manager::FindReadySources(ReadySources* ready)
+			// up inserting duplicates. A source with a zero timeout that was not
+			// selected as the timeout_src can be safely added, whether it's time
+			// to poll or not though.
+			if ( next == 0 && (! time_to_poll || iosource != timeout_src) )
```
I'll put the Poll() fix for timeout_src and timeout == 0 into #2866. I think I have an idea.

> I like the idea of having the possibility to combine both. 

After staring, it should actually be possible to do so.

J-Gras replied:
I might be missing something but I don't see this reflected in #2866 as well. Imho, at least documenting the implicit assumption would be curial.

awelzel replied:
> An IO source can actually be using timeouts and be FD based. It may trigger double processing currently, though that seem least problematic.

Maybe as a bit of a follow-up, because I was rather brief. To me IO sources mixing FD and timeouts seem mostly uncharted territory. My latest read was that they should work out mostly now, though `Process()` possibly called twice when the FD is ready, they returned a timeout of 0.0, it being time_to_poll and there being another IO source with a timeout of 0.0 that was actually selected as the timeout_src for Poll(). And I'd leave it open that we prevent that in the future so they are processed just once.

IMO, once we find limitations and things that don't work quite as needed or expected, should improve from there and if things appear consistent and reasonable, have tests and document it (the above isn't quite there yet I think).

There's also some behaviors that I find rather surprising or questionable. Most prominently, reading a pcap without `suspend_processing()` that has less than 100 packets, FD based IO sources won't ever get polled or processed within the main-loop. With more packets available, FD based IO source processing happens only every 100 packets and in live operating mode 10 packets - which may or may not be the right behavior in certain use-cases.

timwoj commented on src/RunState.cc:
```diff
@@ -319,21 +357,17 @@ void run_loop()
-			update_network_time(util::current_time());
-			expire_timers();
-			}
+		if ( ! time_updated )
```
It'd take a couple extra hits for calling `current_time`. On POSIX systems that should be reasonably fast, but on Windows it's calling `std::chrono` which isn't exactly fast.

timwoj commented on src/Event.cc:
```diff
@@ -189,22 +189,12 @@ void EventMgr::Describe(ODesc* d) const
 
 void EventMgr::Process()
 	{
-	// If we don't have a source, or the source is closed, or we're
```
The reason why this is an IO source is so that we can add the file descriptors from the flares to the kqueue.

awelzel commented on src/iosource/Manager.cc:
```diff
@@ -157,25 +157,24 @@ void Manager::FindReadySources(ReadySources* ready)
+				{
+				ready->push_back({iosource, -1, 0});
+				}
+			else if ( iosource == pkt_src )
```
Would be done in #2866 , but refrained because it doesn't make it look easier.

awelzel commented on src/iosource/PktSrc.h:
```diff
@@ -88,6 +88,21 @@ class PktSrc : public IOSource
+	 * wallclock interval.
+	 *
+	 * The default implementation looks at failing ExtractNextPacket() calls
+	 * and keeps the wallclock timestamp when the source has become idle.
```
Nice, thanks!

awelzel commented:
I've rebased this and added NEWS paragraphs about what I understand has changed. Some may read drastic, but given that barely any existing tests required updates, seems some things were mostly un(der)specified (they aren't really specified now, but baselines do encoded more behavior today).

timwoj commented on NEWS:
```diff
@@ -91,6 +91,11 @@ New Functionality
 
+- Zeek can be prevented from updating ``network_time()`` to the current time
+  by setting ``allow_network_time_forward=F``. Together with ``set_network_time()``
+  or a custom plugin, this allows to control ``network_time()`` without Zeek
```
```suggestion
  or a custom plugin, this allows control of ``network_time()`` without Zeek
```

timwoj reviewed (approved):
Just one nit on the wording in the NEWS.

awelzel commented on src/broker/Manager.cc:
```diff
@@ -1140,11 +1140,6 @@ void Manager::DispatchMessage(const broker::topic& topic, broker::data msg)
 void Manager::Process()
 	{
-	// Ensure that time gets update before processing broker messages, or events
-	// based on them might get scheduled wrong.
```
My understanding: Any script side event handling based on broker messages is done in `event_mgr.Drain()` (after a packet or in the main-loop) and we'll forward time over there if needed.

If now broker working in "real-time" with it's own clock starts acting surprising when Zeek is behind, we should rather see how to fix this on the broker side. Don't think the fast-forward network time on the Zeek side is great.

awelzel replied:
Possibly, yes, though I'm actually thinking more of the inverse: That we maybe need to do a `broker_mgr->AdvanceTime(run_state::network_time)` here as to inform broker about the current time when running in `zeek -r` mode, but I'm treating this separate for now.

:crossed_fingers:

awelzel commented on src/broker/Manager.cc:
```diff
@@ -1221,10 +1216,16 @@ void Manager::Process()
+			//
+			// Do not do this when allow_network_time_forard is set to F
+			// with the assumption that this is unwanted behavior.
+			if ( get_option("allow_network_time_forward")->AsBool() )
+				run_state::detail::update_network_time(util::current_time());
```
Possibly, yes. I'm going to leave this as is for now: It may actually not be needed anymore as we'll have done the time update in the main-loop if there wasn't actually a packet on a worker (and would've done it on a non-worker) before the first `event_mgr.Drain()`.

I might just remove it in a follow-up directly.

awelzel commented on src/RunState.cc:
```diff
@@ -137,6 +137,44 @@ void update_network_time(double new_network_time)
+	return BifConst::allow_network_time_forward;
+	}
+
+static bool maybe_forward_network_time()
```
I think I'll just go with `forward_network_time_if_applicable`. It's `static`, so if something better comes along can easily be changed.

Googling for "wall time" and also "wallclock time" is about elasped time. I'll update NEWS and docs to say "current time" instead... :-/

> Elapsed real time, real time, wall-clock time, wall time, or walltime is the actual time taken from the start of a to the end.

https://en.wikipedia.org/wiki/Elapsed_real_time

J-Gras commented on src/iosource/Manager.cc:
```diff
@@ -157,25 +157,24 @@ void Manager::FindReadySources(ReadySources* ready)
 			double next = iosource->GetNextTimeout();
-			bool added = false;
 
 			if ( timeout == -1 || (next >= 0.0 && next < timeout) )
```
I don't see this addressed in #2866. Is this an oversight or a deliberate choice?

awelzel replied:
Commenting was done a bit in #2866 , but yeah, otherwise adding additional comments hasn't been addressed.
